linera-execution/src/evm/database.rs

```diff
@@ -209,17 +211,54 @@ where
 {
     type Error = ExecutionError;
 
+    /// The `basic_ref` is the function for reading the state of the application.
     fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, ExecutionError> {
+        if address == FAUCET_ADDRESS {
+            return Ok(Some(AccountInfo {
+                balance: FAUCET_BALANCE,
+                ..AccountInfo::default()
+            }));
+        }
         if !self.changes.is_empty() {
-            let account = self.changes.get(&address).unwrap();
-            return Ok(Some(account.info.clone()));
+            // This is the case of service calls with empty storage.
+            let account = self.changes.get(&address);
+            return Ok(account.map(|account| account.info.clone()));
         }
         let mut runtime = self.runtime.lock().unwrap();
+        let account_owner = address.into();
+        // The balances being used are the ones of Linera. So, we need to
+        // access them at first.
+        let balance = runtime.read_owner_balance(account_owner)?;
+
+        let balance: U256 = balance.into();
         let key_info = Self::get_address_key(KeyCategory::AccountInfo as u8, address);
         let promise = runtime.read_value_bytes_new(key_info)?;
         let result = runtime.read_value_bytes_wait(&promise)?;
-        let account_info = from_bytes_option::<AccountInfo>(&result)?;
-        Ok(account_info)
+        let mut account_info = match result {
+            None => AccountInfo::default(),
+            Some(bytes) => bcs::from_bytes(&bytes)?,
+        };
+        // The funds have been immediately deposited in deposit_funds.
+        // The EVM will do the same before even the execution of
+        // the constructor or function.
+        // Therefore, we need to adjust the values.
+        // This will ensure that at any time the balances in EVM
+        // and Linera are exactly matching during the execution.
+        let start_balance = if self.caller == address {
+            balance + self.value
+        } else if self.contract_address == address {
+            assert!(
+                balance >= self.value,
+                "We should have balance >= self.value"
+            );
+            balance - self.value
+        } else {
+            balance
+        };
```

Comment:
(So we return the value _before_ the deposit, even though (in the EVM state) they have already been updated?)